operational_analysis/methods/plant_analysis.py

```diff
@@ -591,75 +695,99 @@ def filter_outliers(self, reanal, outlier_thresh, comb_loss_thresh):
     def run_regression(self, n):
         """
-        Run robust linear regression between Monte-Carlo generated monthly gross energy and wind speed
-        Return Monte-Carlo sampled slope and intercept values (based on their covariance) and report
-        the number of outliers based on the robust linear regression result.
+        Run robust linear regression between Monte-Carlo generated monthly/daily gross energy, 
+        wind speed, temperature and wind direction (if used)
 
         Args:
             n(:obj:`int`): The Monte Carlo iteration number
 
         Returns:
-            :obj:`float`: Monte-carlo sampled slope
-            :obj:`float`: Monte-carlo sampled intercept
+            :obj:`float`: Monte-carlo sampled slope (for linear regression)
+            :obj:`float`: Monte-carlo sampled intercept (for linear regression)
+            :obj:`?`: trained ML model (for ML regression)
         """
         reg_data = self.set_regression_data(n)  # Get regression data
-
-        p, V = np.polyfit(reg_data[0], reg_data[1], 1,
-                          cov=True)  # Perform linear regression, return slope, intercept, and variances
-        # Generate MC-sampled slope and intercept from regression relationship
-        mc_slope, mc_intercept = np.random.multivariate_normal(p, V,
-                                                               1).T
-
+        
+        # Randomly select 80% of the data to perform regression and incorporate some regression uncertainty
+        reg_data = np.array(reg_data.sample(frac = 0.8))
+        
         # Update Monte Carlo tracker fields
-        self._mc_num_points[n] = len(reg_data[0])
-        self._mc_slope[n] = np.float(mc_slope)
-        self._mc_intercept[n] = np.float(mc_intercept)
+        self._mc_num_points[n] = np.shape(reg_data)[0]
+        
+        # Linear regression
+        if self.reg_model == 'lin':
+            reg = LinearRegression().fit(np.array(reg_data[:,0:-1]), reg_data[:,-1])
+            predicted_y = reg.predict(np.array(reg_data[:,0:-1]))
 
-        # Return slope and intercept values
-        return np.float(mc_slope), np.float(mc_intercept)
+            self._mc_slope[n,:] = reg.coef_
+            self._mc_intercept[n] = np.float(reg.intercept_)
+            
+            self._r2_score[n] = r2_score(reg_data[:,-1], predicted_y)
+            self._mse_score[n] = mean_squared_error(reg_data[:,-1], predicted_y)
+            return reg.coef_, reg.intercept_
+        # Machine learning models
+        else: 
+            ml = MachineLearningSetup(self.reg_model)
+            ml.hyper_optimize(np.array(reg_data[:,0:-1]), reg_data[:,-1], n_iter_search = 5, report = False, cv = KFold(n_splits = 2))
+            predicted_y = ml.random_search.predict(np.array(reg_data[:,0:-1]))
+                 
+            self._r2_score[n] = r2_score(reg_data[:,-1], predicted_y)
+            self._mse_score[n] = mean_squared_error(reg_data[:,-1], predicted_y)
```

Comment:
What if the user wants to pass some parameters to the ML algorithm here? Or choose a different one?

I understand it wasn't in your original scope of work - but just adding some variables (that are passed through to here) might make this class more useful and extendable. As one example, you could pass through "keyword arguments" - also known as "kwargs" to MachineLearningSetup. This would provide the user of the AEP method with more control over the MachineLearningSetup object:

up top in __init__:
```
__init__(..., ml_setup_kwargs={})
self.ml_setup_kwargs = ml_setup_kwargs
```
and then here:
```
ml = MachineLearningSetup(self.reg_model, **self.ml_setup_kwargs)
```